source/solvers/mf_navier_stokes.cc

```diff
@@ -786,129 +931,289 @@ MFNavierStokesPreconditionGMG<dim>::initialize(
-      .mg_coarse_grid_reduce;
-  this->coarse_grid_solver_control = std::make_shared<ReductionControl>(
-    max_iterations, tolerance, reduce, false, false);
-  SolverGMRES<VectorType>::AdditionalData solver_parameters;
-  solver_parameters.max_n_tmp_vectors =
-    this->simulation_parameters.linear_solver.at(PhysicsID::fluid_dynamics)
-      .mg_coarse_grid_max_krylov_vectors;
-
-  this->coarse_grid_solver =
-    std::make_shared<SolverGMRES<VectorType>>(*this->coarse_grid_solver_control,
-                                              solver_parameters);
-
   if (this->simulation_parameters.linear_solver.at(PhysicsID::fluid_dynamics)
-        .mg_coarse_grid_preconditioner ==
-      Parameters::LinearSolver::PreconditionerType::amg)
+        .mg_coarse_grid_solver ==
+      Parameters::LinearSolver::CoarseGridSolverType::gmres)
     {
-      TrilinosWrappers::PreconditionAMG::AdditionalData amg_data;
-      amg_data.elliptic = false;
-      if (this->dof_handler.get_fe().degree > 1)
-        amg_data.higher_order_elements = true;
-      amg_data.n_cycles =
+      const int max_iterations =
         this->simulation_parameters.linear_solver.at(PhysicsID::fluid_dynamics)
-          .amg_n_cycles;
-      amg_data.w_cycle =
+          .mg_gmres_max_iterations;
+      const double tolerance =
         this->simulation_parameters.linear_solver.at(PhysicsID::fluid_dynamics)
-          .amg_w_cycles;
-      amg_data.aggregation_threshold =
+          .mg_gmres_tolerance;
+      const double reduce =
         this->simulation_parameters.linear_solver.at(PhysicsID::fluid_dynamics)
-          .amg_aggregation_threshold;
-      amg_data.smoother_sweeps =
+          .mg_gmres_reduce;
+      this->coarse_grid_solver_control = std::make_shared<ReductionControl>(
+        max_iterations, tolerance, reduce, false, false);
+      SolverGMRES<VectorType>::AdditionalData solver_parameters;
+      solver_parameters.max_n_tmp_vectors =
         this->simulation_parameters.linear_solver.at(PhysicsID::fluid_dynamics)
-          .amg_smoother_sweeps;
-      amg_data.smoother_overlap =
-        this->simulation_parameters.linear_solver.at(PhysicsID::fluid_dynamics)
-          .amg_smoother_overlap;
-      amg_data.output_details = false;
-      amg_data.smoother_type  = "ILU";
-      amg_data.coarse_type    = "ILU";
+          .mg_gmres_max_krylov_vectors;
+
+      this->coarse_grid_solver = std::make_shared<SolverGMRES<VectorType>>(
+        *this->coarse_grid_solver_control, solver_parameters);
 
-      std::vector<std::vector<bool>> constant_modes;
-      ComponentMask                  components(dim + 1, true);
       if (this->simulation_parameters.linear_solver
             .at(PhysicsID::fluid_dynamics)
-            .preconditioner ==
-          Parameters::LinearSolver::PreconditionerType::lsmg)
-
+            .mg_gmres_preconditioner ==
+          Parameters::LinearSolver::PreconditionerType::amg)
```

Comment:
I would suggest to move 

https://github.com/chaos-polymtl/lethe/blob/9a43846f67bbc159c7a63fd604884c73625f4ae2/source/solvers/mf_navier_stokes.cc#L962-L1060

and

https://github.com/chaos-polymtl/lethe/blob/9a43846f67bbc159c7a63fd604884c73625f4ae2/source/solvers/mf_navier_stokes.cc#L1111-L1205

to a private function that is shared. Similarly, the logic for ILU can be shared :+1: